l10n_ch_qr_bill_scan/wizard/account_invoice_import.py

```
@@ -0,0 +1,159 @@
+# Copyright 2020 Camptocamp
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import logging
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+from odoo.modules.module import get_resource_path
+
+from ..tools import QR
+
+logger = logging.getLogger(__name__)
+
+try:
+    import cv2
+    import numpy as np
+    import pdf2image
+    from pyzbar.pyzbar import ZBarSymbol, decode
+except ImportError:
+    # Necessary libraries to decode QR from pdf are
+    # not installed
+    logger.warning(
```

I think we should raise an error here because it is important for the QR scan to work properly

Test cases are failing because we cannot extract the partner from the PDF file without these modules.

---

why not add them to test-requirements.txt then?

---

Even though the package was installed successfully, we still require [these libraries](https://github.com/sonhd91/l10n-switzerland/pull/17/commits/b12b3649743dfb7b397e039780e169016426a467#diff-eb0aab1a9ec98c524c9a1268e5496b0104dde25bc901ae4c6e7df278ccc98cf5R5-R6) for it to function it properly. If we don't include them, test case would fail, could we add these lib to test-requirement file?

---

can you add this in the manifest:
```
    "external_dependencies": {
        "deb": ["libzbar0"],
        "python": ["pyzbar", "pdf2image", "opencv-python", "numpy"],
    },
```
libzbar0 is needed for pyzbar indeed
but I don't think that `poppler-utils libzbar-dev zbar-tools` are still needed

---

I tried remove `poppler-utils` and got this error on test case `test_import_QR_pdf`

```
Traceback (most recent call last):
  File "/home/sonhd/.local/share/virtualenvs/odoo16_venv/lib/python3.8/site-packages/pdf2image/pdf2image.py", line 568, in pdfinfo_from_path
    proc = Popen(command, env=env, stdout=PIPE, stderr=PIPE)
  File "/usr/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'pdfinfo'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sonhd/code/odoo16/odoo/project/l10n-switzerland/l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py", line 260, in test_import_QR_pdf
    invoice = self.import_invoice_file(invoice_fp, "qr-bill.pdf")
  File "/home/sonhd/code/odoo16/odoo/project/l10n-switzerland/l10n_ch_qr_bill_scan/tests/test_scan_qrbill.py", line 109, in import_invoice_file
    res = wiz.import_invoice()
  File "/home/sonhd/code/odoo16/odoo/project/edi/account_invoice_import/wizard/account_invoice_import.py", line 788, in import_invoice
    parsed_inv = self.get_parsed_invoice()
  File "/home/sonhd/code/odoo16/odoo/project/l10n-switzerland/l10n_ch_qr_bill_scan/wizard/account_invoice_import.py", line 51, in get_parsed_invoice
    return super().get_parsed_invoice()
  File "/home/sonhd/code/odoo16/odoo/project/edi/account_invoice_import/wizard/account_invoice_import.py", line 647, in get_parsed_invoice
    return self.parse_invoice(self.invoice_file, self.invoice_filename)
  File "/home/sonhd/code/odoo16/odoo/project/edi/account_invoice_import/wizard/account_invoice_import.py", line 510, in parse_invoice
    parsed_inv = self.parse_pdf_invoice(file_data)
  File "/home/sonhd/code/odoo16/odoo/project/l10n-switzerland/l10n_ch_qr_bill_scan/wizard/account_invoice_import.py", line 148, in parse_pdf_invoice
    pdf_img = pdf2image.convert_from_bytes(file_data)[0].convert("RGB")
  File "/home/sonhd/.local/share/virtualenvs/odoo16_venv/lib/python3.8/site-packages/pdf2image/pdf2image.py", line 358, in convert_from_bytes
    return convert_from_path(
  File "/home/sonhd/.local/share/virtualenvs/odoo16_venv/lib/python3.8/site-packages/pdf2image/pdf2image.py", line 127, in convert_from_path
    page_count = pdfinfo_from_path(
  File "/home/sonhd/.local/share/virtualenvs/odoo16_venv/lib/python3.8/site-packages/pdf2image/pdf2image.py", line 594, in pdfinfo_from_path
    raise PDFInfoNotInstalledError(
pdf2image.exceptions.PDFInfoNotInstalledError: Unable to get page count. Is poppler installed and in PATH?
```

---

ok my bad, `poppler-utils` needed then

---

IMHO, Son's concern is valid. Error should be risen at the line causes
it, instead of just single `except` for all 4 lines. OTOH, the
instruction about required packages is helpful.

To have both, we can re-raise the error in the `except` section.

---

- I think the good practice is to never raise an error at import level, because it would prevent the whole odoo instance from loading: we just log the error and only the concerned module will fail
- anyway, now that `external_dependencies` has been defined, dep availability will be checked by odoo at install

---

I tried add this to manifest file, but it seems like Travis doesn't install it

```
  "external_dependencies": {
        "deb": ["libzbar0", "poppler-utils"],
        "python": ["pyzbar", "pdf2image", "opencv-python", "numpy"],
    },
```